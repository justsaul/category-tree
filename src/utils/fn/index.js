export const range = (len) => [...Array(len).keys()]
