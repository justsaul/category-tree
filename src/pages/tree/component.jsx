import React from 'react'
import Provider from '../../components/tree/provider'

import { TYPE as TREE_TYPE } from '../../utils/tree'

const TreePage = (props) => (
    <>
        <Provider {...props} type={TREE_TYPE.RECURSIVE}/>
        <Provider {...props} type={TREE_TYPE.ITERATIVE}/>
    </>
)

export default TreePage
