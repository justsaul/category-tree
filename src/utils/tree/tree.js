import { range } from '../fn'

export const TYPE = {
  RECURSIVE: 'recursive',
  ITERATIVE: 'iterative',
}

export class Tree {
  constructor(value) {
    this.value = value
    this.children = []
  }

  addChildren(value) {
    value instanceof Tree
      ? this.children.push(value)
      : this.children.push(new Tree(value))

    return this
  }

  getValue() {
    return this.value
  }

  getChildren() {
    return this.children
  }

  insert(insertable, atNode, tree=this) {
    if (tree.value === atNode.value) {
      tree.addChildren(insertable)
      return true
    }

    if (~tree.children.length) {
      range(tree.children.length).forEach((i) => {
        this.insert(insertable, atNode, tree.children[i])
      })
    }
    return false
  }
}
