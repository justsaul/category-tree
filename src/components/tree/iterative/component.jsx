import React from 'react'
import PropTypes from 'prop-types'

import { Tree } from '../../../utils/tree'
import { range } from '../../../utils/fn'

export class TreeIterative extends React.Component {
    state = {
        tree: this.props.tree
    }

    traverseTree = (tree) => {
        const stack = [{
            depth: 0,
            element: tree
        }]
        let current
        let children
        let depth
        const returnable = []
        // eslint-disable-next-line
        while (current = stack.pop()) {
            depth = current.depth
            current = current.element
            children = current.children
            // eslint-disable-next-line 
            range(children.length).forEach((pos) => {
                stack.push({
                    element: children[pos],
                    depth: depth + 1
                })
            })
            returnable.push({
                current,
                depth
            })
        }

        return returnable
    }

    render() {
        const { tree } = this.state
        const { update } = this.props

        return (
            <div>
                {this.traverseTree(tree).map((iter, i) => {
                    return (
                        <div
                            key={i}
                            style={{
                                paddingLeft: `${(iter.depth + 1) * 20}px`,
                            }}
                        >
                            {iter.current.value}
                            <button
                                onClick={() => { update('element', iter.current)}}
                                title="addOperation"
                            >
                                {'Add operation'}
                            </button>
                        </div>
                    )
                })}
            </div>
        )
    }
}

TreeIterative.propTypes = {
    tree: PropTypes.instanceOf(Tree).isRequired,
    update: PropTypes.func.isRequired,
}
