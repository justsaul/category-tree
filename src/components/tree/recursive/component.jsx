import React from 'react'
import PropTypes from 'prop-types'

import { Tree } from '../../../utils/tree'

export class TreeRecursive extends React.Component {
    state = {
        tree: this.props.tree
    }

    render() {
        const { tree, update } = this.props

        if (tree) {
            return (
                <div style={{paddingLeft: '20px'}}>
                    {tree.value}
                    <button
                        onClick={() => update('element', this.state.tree)}
                        title="addOperation"
                    >
                        {'Add operation'}
                    </button>
                    {
                        tree.children.map((desc, i) => {
                            return (
                                <TreeRecursive
                                    tree={desc}
                                    key={tree.value + i}
                                    update={update}
                                />
                            )
                        })
                    }
                </div>
            )
        } else {
            return (
                <div></div>
            )
        }
    }
}

TreeRecursive.propTypes = {
    tree: PropTypes.instanceOf(Tree).isRequired,
    update: PropTypes.func.isRequired,
}