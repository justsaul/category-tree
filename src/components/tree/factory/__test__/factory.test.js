import React from 'react'
import { shallow } from 'enzyme'

import { TYPE as TREE_TYPE } from '../../../../utils/tree'
import { Tree } from '../../../../utils/tree'
import { TreeIterative } from '../../iterative'
import { TreeRecursive } from '../../recursive'
import { TreeFactory } from '../'

describe('<TreeFactory/>', () => {
    it('should render without exploding', () => {
        const Lambda = new Tree('Lambda')
        const wrapper = shallow(
            <TreeFactory type={TREE_TYPE.ITERATIVE} tree={Lambda} update={jest.fn()}/>
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('should provide only <TreeIterative/> tree based on type', () => {
        const Lambda = new Tree('Lambda')
        Lambda.addChildren(new Tree('Children'))
        
        const wrapper = shallow(
            <TreeFactory type={TREE_TYPE.ITERATIVE} tree={Lambda} update={jest.fn()}/>
        )

        expect(wrapper.find(TreeIterative).length).toBe(1)
        expect(wrapper.find(TreeRecursive).length).toBe(0)
    })

    it('should provide only <TreeRecursive/> tree based on type', () => {
        const Lambda = new Tree('Lambda')
        Lambda.addChildren(new Tree('Children'))
        
        const wrapper = shallow(
            <TreeFactory type={TREE_TYPE.RECURSIVE} tree={Lambda} update={jest.fn()}/>
        )

        expect(wrapper.find(TreeRecursive).length).toBe(1)
        expect(wrapper.find(TreeIterative).length).toBe(0)
    })
})
