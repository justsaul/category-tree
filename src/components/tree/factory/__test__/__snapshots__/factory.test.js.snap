// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TreeFactory/> should render without exploding 1`] = `ShallowWrapper {}`;
